### PR TITLE
Initialize unit tests for health service/endpoiont

### PR DESCRIPTION
**Pull request by tfussell**
The health endpoint test mocks the k8s client to calculate cluster health. I think this will be essential as we develop the health logic, but for now it can't be fleshed out until we actually implement the health check logic.
The health service test simply ensures that `Config` contains all of the required fields. I'm not too happy about having to check all of these individually since it feels like the type system should help with this, but maybe it's just a Go thing.

Towards https://github.com/giantswarm/giantswarm/issues/6893

### CODE REVIEW

File: server/endpoint/health/searcher/endpoint_test.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	"github.com/spf13/viper"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	"github.com/giantswarm/health-service/flag"
+	"github.com/giantswarm/health-service/server/middleware"
+	"github.com/giantswarm/health-service/service"
+)
+
+var testCases = []struct {
```
Comment (MarcelMue) [inline server/endpoint/health/searcher/endpoint_test.go]: I would like to have this more aligned with `fmt`.
This structure looks too unique to me in general - can we make it fit better?

https://github.com/giantswarm/fmt/blob/master/go/unit_tests.md#table-driven-tests

File: service/health/searcher/service_test.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	"github.com/giantswarm/micrologger/microloggertest"
+)
+
+func Test_Service(t *testing.T) {
+	testCases := []struct {
+		RequestFactory  func() Request
+		ExpectedFactory func() Response
+		ErrorMatcher    func(error) bool
+	}{
+		// Case 1. Having an empty request given there should be one master with
```
Comment (MarcelMue) [inline service/health/searcher/service_test.go]: Same here, we usually have the `name` field and so on to make it easier to handle these tests.

File: server/endpoint/health/searcher/endpoint_test.go (hunk 1 above)
Comment (tfussell) [reply server/endpoint/health/searcher/endpoint_test.go]: Ah, I hadn't seen those docs. I'll get this cleaned up.

Comment (tfussell) [reply server/endpoint/health/searcher/endpoint_test.go]: This should be fixed in my latest commit. Thanks for telling me about the `fmt` docs regarding unit tests.

File: service/health/searcher/service_test.go (hunk 2 above)
Comment (tfussell) [reply service/health/searcher/service_test.go]: This should be fixed in my latest commit too.

File: server/endpoint/health/searcher/endpoint_test.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+				}
+
+				healthService, err = health.New(healthConfig)
+				if err != nil {
+					t.Fatal("Error creating health service: ", err)
+				}
+			}
+
+			// Create health endpoint.
+			cfg := Config{
```
Comment (MarcelMue) [inline server/endpoint/health/searcher/endpoint_test.go]:
```suggestion
			c := Config{
```

File: server/endpoint/health/searcher/endpoint_test.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+				Logger:     microloggertest.New(),
+				Middleware: &middleware.Middleware{},
+				Service: &service.Service{
+					Health: healthService,
+				},
+			}
+
+			endpoint, err := New(cfg)
```
Comment (MarcelMue) [inline server/endpoint/health/searcher/endpoint_test.go]:
```suggestion
			endpoint, err := New(c)
```

File: server/endpoint/health/searcher/endpoint_test.go (hunk 3 above)
Comment (MarcelMue) [reply server/endpoint/health/searcher/endpoint_test.go]: Would be even nicer with the whole 
```
var endpoint ... 
{
   c := Config{
  ...
  }
  endpoint = New(c)
}

File: service/health/searcher/service.go [hunk 5]
```diff
@@ -29,6 +38,14 @@ func New(config Config) (*Service, error) {
 	if config.Provider == "" {
 		return nil, microerror.Maskf(invalidConfigError, "provider must not be empty")
 	}
+	providers := []string{
```
Comment (MarcelMue) [inline service/health/searcher/service.go]: This should be a `const` - or maybe in a `key` function?

Comment (tfussell) [reply service/health/searcher/service.go]: Moved to key.

File: server/endpoint/health/searcher/endpoint_test.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
...bernetes.io/os":"linux","giantswarm.io/provider":"kvm","ip":"172.23.0.242","kubernetes.io/arch":"amd64","kubernetes.io/hostname":"worker-oh8xv-c549d84f-24z2s","kubernetes.io/os":"linux","kubernetes.io/role":"worker","kvm-operator.giantswarm.io/version":"3.5.0","node-role.kubernetes.io/worker":"","node.kubernetes.io/worker":"","role":"worker"},"lastTransitionTime":"2019-09-23T17:55:06.434470124Z","name":"worker-oh8xv-c549d84f-24z2s","version":"3.5.0"},{"labels":{"beta.kubernetes.io/arch":"amd64","beta.kubernetes.io/os":"linux","giantswarm.io/provider":"kvm","ip":"172.23.0.86","kubernetes.io/arch":"amd64","kubernetes.io/hostname":"worker-pq295-6c4bbcccf7-kql2k","kubernetes.io/os":"linux","kubernetes.io/role":"worker","kvm-operator.giantswarm.io/version":"3.5.0","node-role.kubernetes.io/worker":"","node.kubernetes.io/worker":"","role":"worker"},"lastTransitionTime":"2019-09-23T17:55:06.434470788Z","name":"worker-pq295-6c4bbcccf7-kql2k","version":"3.5.0"}],"resources":null,"scaling":{"desiredCapacity":0},"versions":[{"date":"0001-01-01T00:00:00Z","lastTransitionTime":"2019-09-23T12:01:19.910404547Z","semver":"3.5.0"}]},"kvm":{"nodeIndexes":{"hhn5a":1,"hi5ie":2,"oh8xv":4,"pq295":3}}}}`,
```
Comment (kopiczko) [inline server/endpoint/health/searcher/endpoint_test.go]: Does this response have to be so crazy? I wonder whether this is maintainable. Can it be trimmed a little bit?

File: server/endpoint/health/searcher/endpoint_test.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+			}
+
+			g8sClient, err := versioned.NewForConfig(restConfig)
+			if err != nil {
+				t.Fatal("expected", nil, "got", err)
+			}
+
+			f := flag.New()
+			v := viper.New()
+			v.Set(f.Service.Provider.Kind, "kvm")
```
Comment (kopiczko) [inline server/endpoint/health/searcher/endpoint_test.go]: It feels to me like `"kvm"` should be test input and we can test for multiple providers.

File: server/endpoint/health/searcher/endpoint_test.go [hunk 8]
```diff
@@ -33,7 +33,7 @@ func Test_Health_Endpoint(t *testing.T) {
...
...bernetes.io/os":"linux","giantswarm.io/provider":"kvm","ip":"172.23.0.242","kubernetes.io/arch":"amd64","kubernetes.io/hostname":"worker-oh8xv-c549d84f-24z2s","kubernetes.io/os":"linux","kubernetes.io/role":"worker","kvm-operator.giantswarm.io/version":"3.5.0","node-role.kubernetes.io/worker":"","node.kubernetes.io/worker":"","role":"worker"},"lastTransitionTime":"2019-09-23T17:55:06.434470124Z","name":"worker-oh8xv-c549d84f-24z2s","version":"3.5.0"},{"labels":{"beta.kubernetes.io/arch":"amd64","beta.kubernetes.io/os":"linux","giantswarm.io/provider":"kvm","ip":"172.23.0.86","kubernetes.io/arch":"amd64","kubernetes.io/hostname":"worker-pq295-6c4bbcccf7-kql2k","kubernetes.io/os":"linux","kubernetes.io/role":"worker","kvm-operator.giantswarm.io/version":"3.5.0","node-role.kubernetes.io/worker":"","node.kubernetes.io/worker":"","role":"worker"},"lastTransitionTime":"2019-09-23T17:55:06.434470788Z","name":"worker-pq295-6c4bbcccf7-kql2k","version":"3.5.0"}],"resources":null,"scaling":{"desiredCapacity":0},"versions":[{"date":"0001-01-01T00:00:00Z","lastTransitionTime":"2019-09-23T12:01:19.910404547Z","semver":"3.5.0"}]},"kvm":{"nodeIndexes":{"hhn5a":1,"hi5ie":2,"oh8xv":4,"pq295":3}}}}`,
```
Comment (MarcelMue) [inline server/endpoint/health/searcher/endpoint_test.go]: Let's discuss the general idea at another point but the formatting should change for sure.
We sometimes put these things into extra files as a const - I am not sure about the usefulness of such a specific object inside the test.

Changes to the spec of the kvmConfig could have unforeseen / unfortunate consequences here. I am happy to be wrong though.

File: service/health/key/key.go [hunk 9]
```diff
@@ -0,0 +1,12 @@
+package key
+
+var Providers = [...]string{"aws", "azure", "kvm"}
+
+func IsKnownProvider(value string) bool {
```
Comment (MarcelMue) [inline service/health/key/key.go]: I like it.

File: server/endpoint/health/searcher/endpoint_test.go [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+			}
+
+			req, err := http.NewRequest(Method, Path, nil)
+			if err != nil {
+				t.Fatal("Unexpected HTTP error:", err)
+			}
+
+			ctx := req.Context()
+			type contextKey string
+			ctx = context.WithValue(ctx, contextKey("Content-Type"), "application/json; charset=utf-8")
```
Comment (kopiczko) [inline server/endpoint/health/searcher/endpoint_test.go]: Is it required to create the context like that? Can't it be simply `context.Background()`? (I didn't check the code 🙈 )

File: service/health/searcher/kvm.go [hunk 11]
```diff
@@ -2,12 +2,30 @@ package searcher
...
 import (
 	"context"
+
+	"github.com/giantswarm/microerror"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 // searchKVMCR searches for the cluster config in AWSClusterConfigs resources.
 func (s *Service) searchKVMCR(ctx context.Context, request Request) (*Response, error) {
+	kvmCR, err := s.G8sClient.ProviderV1alpha1().KVMConfigs("default").Get(request.ClusterID, v1.GetOptions{})
```
Comment (MarcelMue) [inline service/health/searcher/kvm.go]: Let's do this in a separate PR and get the basic unit tests merged first.

File: server/endpoint/health/searcher/endpoint_test.go [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+				t.Fatalf("error == %#v, want nil", err)
+			case err == nil && tc.errorMatcher != nil:
+				t.Fatalf("error == nil, want non-nil")
+			case !tc.errorMatcher(err):
+				t.Fatalf("error == %#v, want matching", err)
+			}
+
+			endpointResponseTyped, ok := endpointResponse.(Response)
+			if !ok {
+				t.Fatal("Unexpected response type")
```
Comment (kopiczko) [inline server/endpoint/health/searcher/endpoint_test.go]:
```suggestion
				t.Fatal("endpointResponse.(type) = %T, want %T", endpointResponse , endpointResponseTyped)
```

File: service/health/searcher/service_test.go [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+import (
+	"strconv"
+	"testing"
+
+	apiextensionsfake "github.com/giantswarm/apiextensions/pkg/clientset/versioned/fake"
+	"github.com/giantswarm/micrologger/microloggertest"
+)
+
+func Test_Health_New(t *testing.T) {
```
Comment (kopiczko) [inline service/health/searcher/service_test.go]: I'd skip this test. IMO testing `IsInvalidConfig` introduces more clutter than value.

File: server/endpoint/health/searcher/endpoint_test.go (hunk 8 above)
Comment (tfussell) [reply server/endpoint/health/searcher/endpoint_test.go]: On the one hand, I agree that this could be fragile. On the other hand, the whole health determination is based on this structure. If we change `KVMConfig`, we'll have to update this service too so I think they're necessarily tightly coupled. Open to suggestions for a better/more maintainable place to keep this data though.

File: service/health/searcher/kvm.go (hunk 11 above)
Comment (tfussell) [reply service/health/searcher/kvm.go]: Okay, will branch and revert.

Comment (tfussell) [reply service/health/searcher/kvm.go]: Moved to https://github.com/giantswarm/health-service/tree/kvm-health

File: server/endpoint/health/searcher/endpoint_test.go (hunk 8 above)
Comment (tfussell) [reply server/endpoint/health/searcher/endpoint_test.go]: I reverted this. We'll figure it out once we're implementing health check logic.

File: service/health/searcher/service_test.go (hunk 13 above)
Comment (MarcelMue) [reply service/health/searcher/service_test.go]: I am personally fine with having them here, as we won't have too many services in this microservice.
So if anything I think we can keep it here and still get rid of it later if it gets annoying.

File: server/endpoint/health/searcher/endpoint_test.go (hunk 6 above)
Comment (tfussell) [reply server/endpoint/health/searcher/endpoint_test.go]: I thought it might be nice to have a "real" response taken from `kubectl` while we're developing the health logic. I've removed it for now as it's making the PR too large.

File: server/endpoint/health/searcher/endpoint_test.go (hunk 10 above)
Comment (tfussell) [reply server/endpoint/health/searcher/endpoint_test.go]: I think you're right. And this endpoint doesn't even use the request, only pulls the id from the path using mux. Fixed!

File: server/endpoint/health/searcher/endpoint_test.go (hunk 7 above)
Comment (tfussell) [reply server/endpoint/health/searcher/endpoint_test.go]: Done.

File: server/endpoint/health/searcher/endpoint_test.go (hunk 12 above)
Comment (tfussell) [reply server/endpoint/health/searcher/endpoint_test.go]: Done.

File: service/health/searcher/service_test.go (hunk 13 above)
Comment (tfussell) [reply service/health/searcher/service_test.go]: It currently provides a small amount of value and can be extended in the future if we come up with more tests so I think I might keep it and, as Marcel said, we can get rid of it in the future if it seems useless.

**Review by MarcelMue — APPROVED** [review]
